subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -115,180 +110,101 @@ public ClassSetDiff findChangesSince(ClassSetAnalysis other) {
+     * If the classes had any changes to inlineable constants, these need to be provided as the second parameter.
      */
     public DependentsSet findTransitiveDependents(Collection<String> classes, Map<String, IntSet> constants) {
         if (classes.isEmpty()) {
             return DependentsSet.empty();
         }
-        final Set<String> accessibleResultClasses = new HashSet<>();
-        final Set<String> privateResultClasses = new HashSet<>();
-        final Set<GeneratedResource> resultResources = new HashSet<>();
-        for (String cls : classes) {
+        String fullRebuildCause = annotationProcessingData.getFullRebuildCause();
+        if (fullRebuildCause != null) {
+            return DependentsSet.dependencyToAll(fullRebuildCause);
+        }
+        final Set<String> privateDependents = new HashSet<>();
+        final Set<String> accessibleDependents = new HashSet<>(annotationProcessingData.getGeneratedTypesDependingOnAllOthers());
+        final Set<GeneratedResource> dependentResources = new HashSet<>(annotationProcessingData.getGeneratedResourcesDependingOnAllOthers());
+        for (String cls : Iterables.concat(classes, annotationProcessingData.getGeneratedTypesDependingOnAllOthers())) {
             DependentsSet d = findTransitiveDependents(cls, constants.getOrDefault(cls, IntSets.emptySet()));
             if (d.isDependencyToAll()) {
                 return d;
             }
-            if (d.isEmpty()) {
-                continue;
-            }
-            Set<String> accessibleDependentClasses = d.getAccessibleDependentClasses();
-            Set<String> privateDependentClasses = d.getPrivateDependentClasses();
-            Set<GeneratedResource> dependentResources = d.getDependentResources();
 
-            accessibleResultClasses.addAll(accessibleDependentClasses);
-            privateResultClasses.addAll(privateDependentClasses);
-            resultResources.addAll(dependentResources);
+            accessibleDependents.addAll(d.getAccessibleDependentClasses());
+            privateDependents.addAll(d.getPrivateDependentClasses());
+            dependentResources.addAll(d.getDependentResources());
         }
-        return DependentsSet.dependents(privateResultClasses, accessibleResultClasses, resultResources);
+        return DependentsSet.dependents(privateDependents, accessibleDependents, dependentResources);
     }
 
-    /**
-     * Computes the transitive dependents of a changed class.
-     * If the class had any changes to inlineable constants, these need to be provided as the second parameter.
-     */
-    public DependentsSet findTransitiveDependents(String className, IntSet constants) {
-        String fullRebuildCause = annotationProcessingData.getFullRebuildCause();
-        if (fullRebuildCause != null) {
-            return DependentsSet.dependencyToAll(fullRebuildCause);
-        }
-
-        DependentsSet deps = getDependents(className);
```

Comment:
This was unnecessarily complicated, basically unrolling the first iteration of the transitive dependency detection loop.